client/client_pool/include/client/client_pool/client_pool_timer.hpp

```diff
@@ -15,49 +15,45 @@
 
 #include <chrono>
 #include <future>
-#include <stdexcept>
 
-#include <boost/asio/deadline_timer.hpp>
-#include <boost/asio/io_service.hpp>
-#include <boost/date_time/posix_time/posix_time_duration.hpp>
-#include <boost/system/error_code.hpp>
+#include <asio/steady_timer.hpp>
+#include <asio/io_context.hpp>
 
 namespace concord_client_pool {
 template <typename ClientT>
 class Timer {
  public:
   using Clock = std::chrono::high_resolution_clock;
   Timer(std::chrono::milliseconds timeout, std::function<void(ClientT&&)> on_timeout)
-      : timeout_{timeout}, on_timeout_{on_timeout}, timer_(io_service_) {
+      : timeout_{timeout}, on_timeout_{on_timeout}, timer_(io_context_) {
     if (timeout_.count() > 0) {
-      timer_thread_future_ = std::async(std::launch::async, &Timer::WorkerThread, this);
+      timer_thread_future_ = std::async(std::launch::async, &Timer::work, this);
     }
   }
 
-  ~Timer() { stop(); }
+  ~Timer() { stopTimerThread(); }
 
-  void stop() {
+  void stopTimerThread() {
     if (timeout_.count() == 0) {
       return;
     }
-    stop_ = true;
-    io_service_.stop();
+    io_context_.stop();
     timer_thread_future_.wait();
   }
 
-  void set(const ClientT& client) {
-    if (timeout_.count() == 0) {
+  void start(const ClientT& client) {
+    if (timeout_.count() == 0 || not timer_thread_future_.valid() || io_context_.stopped()) {
```

Comment:
I would add a log here.

Reply:
The implementation didn't use logging before and I don't see it urgent to do so. However, I agree with you that a debug log would be helpful. To me this is a future refactoring task as well (together with re-thinking the lifecycle of the timer object, the constructor, dynamic timeout, ...) but I can add it if you feel it adds value now.

Reply:
I saw that you added logging a couple of hours ago. I will add the message.